### PR TITLE
Add the `Frame::sp` method to get the frame's stack pointer

### PR DESCRIPTION
**Pull request by fitzgen**
r? @alexcrichton

### CODE REVIEW

**Comment by fitzgen** [comment]
No idea what is up with [the `arm-unknown-linux-gnueabihf` tests](https://github.com/rust-lang/backtrace-rs/pull/341/checks?check_run_id=744081109#step:6:466):

```
qemu: Unsupported syscall: 397
test many_threads ... ok
test smoke_test_frames ... ok
fatal runtime error: failed to initiate panic, error 9
error: test failed, to rerun pass '--test smoke'

Caused by:
  process didn't exit successfully: `qemu-arm -L /usr/arm-linux-gnueabihf /checkout/target/arm-unknown-linux-gnueabihf/debug/deps/smoke-10e38a2f0d8ebee1` (signal: 6, SIGABRT: process abort signal)
##[error]Process completed with exit code 101
```

File: tests/smoke.rs [hunk 1]
```diff
@@ -228,3 +242,79 @@ fn is_serde() {
...
+        refs: &'a mut Vec<usize>,
+        child_sp: &'a mut Option<usize>,
+        child_ref: &'a mut Option<usize>,
+    ) -> impl FnMut(&backtrace::Frame) -> bool + 'a {
+        move |frame| {
+            eprintln!("\n=== frame ===================================");
+
+            let mut is_recursive_stack_references = false;
+            backtrace::resolve(frame.ip(), |sym| {
+                is_recursive_stack_references |= (LIBBACKTRACE || GIMLI_SYMBOLIZE)
```
Comment (alexcrichton) [inline tests/smoke.rs:286]: Is the `LIBBACKTRACE || GIMLI_SYMBOLIZE` part here necessary? It looks like this handles a non-present name?

File: tests/smoke.rs [hunk 2]
```diff
@@ -228,3 +242,79 @@ fn is_serde() {
...
+
+    // NB: the following `make_*` functions are pulled out of line, rather than
+    // defining their results as inline closures at their call sites, so that
+    // the resulting closures don't have "recursive_stack_references" in their
+    // mangled names.
+
+    fn make_trace_closure<'a>(
+        refs: &'a mut Vec<usize>,
+        child_sp: &'a mut Option<usize>,
+        child_ref: &'a mut Option<usize>,
```
Comment (alexcrichton) [inline tests/smoke.rs:279]: I think these two parameters could be local variables moved into the closure, right? I don't think they need to be defined out-of-line and passed as arguments

File: tests/smoke.rs [hunk 3]
```diff
@@ -4,6 +4,20 @@ use std::thread;
 static LIBBACKTRACE: bool = cfg!(feature = "libbacktrace") && !cfg!(target_os = "fuchsia");
 static GIMLI_SYMBOLIZE: bool = cfg!(all(feature = "gimli-symbolize", unix, target_os = "linux"));
 
+const HAVE_SP_IMPL: bool = cfg!(any(
```
Comment (alexcrichton) [inline tests/smoke.rs:7]: This looks pretty gnarly, I'm curious how you came across this? Ideally we just wouldn't list this at all here and we'd just always run the test?

File: tests/smoke.rs (hunk 1 above)
Comment (fitzgen) [reply tests/smoke.rs:286]: If we can't rely on symbolication, then we can't determine which stack frames are the ones we are interested in (due to inlining happening or not), and therefore don't know when to pop from our stack of on-stack references.

File: tests/smoke.rs (hunk 3 above)
Comment (fitzgen) [reply tests/smoke.rs:7]: I took the `cfg`s from when dbghelp or libunwind are used and `any`'d them. I suppose I could remove this and just check that the returned SP is non-null, which would be easier to maintain.

File: tests/smoke.rs (hunk 2 above)
Comment (fitzgen) [reply tests/smoke.rs:279]: Done.

File: tests/smoke.rs (hunk 3 above)
Comment (fitzgen) [reply tests/smoke.rs:7]: Done.

**Comment by alexcrichton** [comment]
Hm ok so the failing test here seems semi-legitimate. It looks like the test is panicking but in emulation the panic fails for... whatever reason. The test harness is capturing output to print later but the abort means it doesn't get printed. Anyway passing `--nocapture` the output looks like:

```
running 3 tests
test many_threads ... ok
test smoke_test_frames ... symbol  ip:0xffe9a950 address:0xffe9a950 name:smoke::smoke_test_frames::frame_4::hc85e98dee3b06640 file:/checkout/tests/smoke.rs lineno:18 
symbol  ip:0xffe9a908 address:0xffe9a908 name:smoke::smoke_test_frames::frame_3::h6d55155098f3e65e file:/checkout/tests/smoke.rs lineno:15 
symbol  ip:0xffe9a8ec address:0xffe9a8ec name:smoke::smoke_test_frames::frame_2::h98465abd29b4bfc6 file:/checkout/tests/smoke.rs lineno:14 
symbol  ip:0xffe9a8d0 address:0xffe9a8d0 name:smoke::smoke_test_frames::frame_1::hcd16b4600e849d0d file:/checkout/tests/smoke.rs lineno:13 
symbol  ip:0xffe9a8b4 address:0xffe9a8b4 name:smoke::smoke_test_frames::h4b3feb7e02233f3e file:/checkout/tests/smoke.rs lineno:12 
ok
test sp_smoke_test ... 
=== frame ===================================
name = _ZN9backtrace9backtrace9libunwind5trace17he04a1f50969a4bcaE
name = _ZN9backtrace9backtrace20trace_unsynchronized17h414525a2d0b36f54E
sp  = 0x5

=== frame ===================================
name = _ZN9backtrace9backtrace5trace17h4bf49cef1db4e66fE
sp  = 0x5

=== frame ===================================
name = _ZN5smoke13sp_smoke_test26recursive_stack_references17h6922757508a9c368E
sp  = 0x5
ref = 0xffe8ccdc

=== frame ===================================
name = _ZN5smoke13sp_smoke_test26recursive_stack_references17h6922757508a9c368E
sp  = 0x5
ref = 0xffe8cd84
thread 'main' panicked at 'assertion failed: sp >= child_ref', tests/smoke.rs:301:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 9
error: test failed, to rerun pass '--test smoke'

Caused by:
  process didn't exit successfully: `qemu-arm -L /usr/arm-linux-gnueabihf /checkout/target/arm-unknown-linux-gnueabihf/debug/deps/smoke-10e38a2f0d8ebee1 --nocapture` (signal: 6, SIGABRT: process abort signal)
```

**Comment by fitzgen** [comment]
Okay! I've got `arm-unknown-linux-gnueabihf` working locally, I think that was the only one failing, but lets see what CI thinks...

**Comment by alexcrichton** [comment]
👍

**Pull request merged**